DIPs/1NNN-ADR.md

```diff
@@ -0,0 +1,479 @@
+is the lowered druntime template instantiation and the `Argument`s form the remainder of the tuple elements.
+
+The druntime `object._d_interpolated_string` struct constructor or function is called with no runtime arguments and a compile time argument list constructed as follows:
+
+If the `Element` is:
+
+* `Character`, it is written to the output string.
+* `'$$'`, a '$' is written to the output string.
+
+If a `'$'` occurs without a following `'$'`, it terminates the current output string (if any), appends it to the template argument list, and appends a new argument to the template instantiation as follows:
+
+* `'$' Argument`, then `_d_interpolated_format_spec(null)` is appended to the argument list.
+* `'$' '{' FormatString '}' Argument`, then `_d_interpolated_format_spec(FormatString)` is appended to the argument list. Note that `FormatString` may be the empty string, which is passed as `""` instead of as `null`. For example `i"${}foo"` is passed as `_d_interpolated_string!(_d_interpolated_format_spec(""))()`.
+
+If characters remain, it continues the processing loop with a fresh template argument to `_d_interpolated_string`.
+
+The result of this `_d_interpolated_string` function call becomes the first element of the tuple.
+
+If the `Argument` is an `Identifier`, it is appended to the tuple as an `IdentifierExpression`.
+If the `Argument` is an `Expression`, it is lexed and parsed (including the surrounding parentheses)
+(similar to `MixinExpressions`; you could think of the compiler as putting `mixin(...)` around the extracted string) and appended to the tuple as an `Expression`.
+
+Compile-time errors will be generated if the `Elements` do not fit the grammar.
+
+### Diagnostics
+
+An `i""` string is likely not be compatible with existing D `string` functions and variables. The compiler SHOULD provide helpful hints in the error messages to point new users in the right direction to understand how to use the new feature. This message MAY link to a web page to educate the user on which solution is appropriate for them.
+
+```
+void foo(string s) {}
+
+foo(i"test $(4)"); // a CT error due to type mismatch. The compiler should suggest functions (see below) to remedy the user's problem.
+```
+
+### Concatenations
+
+In order to facilitate convenient formatting of long strings, if a `StringLiteral` follows an `InterpolatedString`,
+it is appended to the `InterpolatedString` in the parsing pass. (Not the lexer pass.)
+A `StringPostfix` may not appear on any of them.
+Such concatenations enable the various string literal types to be used as `InterpolatedStrings`,
+such as:
+
+```
+i""q{apples and $("bananas")}
+```
+yielding a tuple expression:
+```
+_d_interpolated_string!("apples and ", _d_interpolated_format_spec(null)), ("bananas")
+```
+
+### Library additions
+
+While the compiler only lowers to library calls and thus does not need to know about any of this, user functionality relies on a small library implementation.
```

Comment:
What the struct/function does is essential to this DIP. While the implementation may change, the end result must be defined so code can reasonably know what to expect.

To that end, I think the easiest thing is for the spec to say 3 things:

1. The struct will have an alias to the Parts.
2. The struct will define a  `toFormatString` accessor, and define it's aspects
3. For the use case of calling C functions, the struct will alias this itself to a null terminated const char *, only when all format specifiers are provided.

Then you can give a sample implementation.

Reply:
oh yeah, and actually the name of the type will have to be public somehow (my personal preference is to have the compiler use an ugly name, then the library can alias it, potentially in a module like `core.interpolation;`) which has the formal public interface.

But public name is important so people can actually use it in their own overloads.